analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereference.cs

```diff
@@ -947,4 +947,26 @@ public static void MixedConstraints()
             }
         }
     }
+
+    // https://github.com/SonarSource/sonar-dotnet/issues/4784
+    class Repro_4784
+    {
+        public static long Reproducer()
+        {
+            long sum = 0;
+            long[] array = new[] { 1L, 2L, 3L };
+            List<long> test = array.ToList();
+            if (test?.Count == 0)
+            {
+                // Do something
+            }
+
+            foreach (long number in test) // Noncompliant
```

Comment:
I'm not sure if this is really a false positive (as the title says). We raise the issue due to the `?` from `if (test?.Count == 0)`. The engine reasoning: if test can be null on this line it can be null also on the `foreach` one.

What do you think about this?

As far as I can tell it will be great to raise on line 959 that `test` cannot be null, but currently by design the engine does not know that `ToList()` will never return null.

Reply:
Please take a look at it, I added some test cases like we discussed and it seems to me that the issue is really related to ToList().